### PR TITLE
Fix ndless_installer_4.5 on macOS

### PR DESCRIPTION
**Pull request by unilock**
Mostly untested. Assumes the following have been installed with brew:

- binutils
- boost
- gcc
- git
- gmp
- libmpc
- mpfr
- wget
- zlib

May break if more than one gcc version is installed.

### CODE REVIEW

**Comment by adriweb** [comment]
I'll try that later as I need to update my toolchain anyway and I previously made some changes about that...
I'm not sure everything in the patch is actually needed. I built the toolchain just fine with clang, which allows to avoid all the gcc version stuff altogether since it'll take the default xcode one.
I believe the only change I previously needed to make was to remove `with-python` from the GDB build options. Which, funnily enough, isn't part of this patch.

File: ndless/src/installer-4.5/Makefile [hunk 1]
```diff
@@ -26,8 +26,13 @@ all: $(DISTDIR)/ndless_installer_4.5.0.tns
 %.bin: %.o
 	$(OBJCOPY) -O binary $^ $@
 
+OS_NAME := $(shell uname -s | tr A-Z a-z)
 payload: ndless_installer.bin
-	(../tools/MakeChunkDispatch/MakeChunkDispatch 0x1800e7f4 0x12480000; cat $^) | base64 -w0 > $@ || rm $@
+	ifeq ($(OS_NAME),darwin)
```
Comment (Vogtinator) [inline ndless/src/installer-4.5/Makefile]: This should be done by using feature detection instead. `base64` doesn't have/need the `-w0` option on some Linux distros either.

**Comment by Vogtinator** [comment]
I don't think the `brew` specific changes make sense. AFAIK `brew` (optionally?) sets up `$PATH`, so it should already find everything correctly.

The `unalias` calls in `./bulid_toolchain.sh` are noops anyway as they only influence the current shell, which is exited directly afterwards.

**Comment by adriweb** [comment]
So, I've been able to use `build_toolchain.sh` just fine on macOS today, I just updated the versions of the deps: https://github.com/ndless-nspire/Ndless/pull/201
I also removed `--with-python` because I had problems in the past with it.

So I guess your modifications aren't actually needed? Must be some configuration on your side, for brew etc. I don't believe I have anything special in my PATH to make it work...

Regarding the base64 thing in installer-4.5, yeah, we need to have a solution properly, as it wouldn't work on macOS otherwise.

Anyway, I agree with what @Vogtinator said, otherwise.

**Comment by unilock** [comment]
@adriweb That's strange; I ran into errors when first attempting to build the toolchain with clang. Though, trying it again with only the changes from #201 applied, I'm able to build it just fine (notably with `--with-python`).
So my changes to build_toolchain.sh were unnecessary.

**Comment by unilock** [comment]
@Vogtinator I _believe_ I've done as you asked; I'm not 100% certain if my modification actually works. Notably, while rebuilding the toolchain on macOS to test my handiwork, I ran into the following issue:
```
Extracting binutils-2.34.tar.bz2...
tar: Option -a is not permitted in mode -x
```
Removing `a` from the `tar` options in `build_toolchain.sh` seems to have fixed this. Not sure why it was there in the first place.

**Comment by adriweb** [comment]
That's probably due to the other recent changes, uh oh :p

**Comment by unilock** [comment]
`ifeq ($(shell base64 -w 2>/dev/null),)` is probably more correct than `ifeq ($(base64 -w 2>/dev/null),)` - but apparently still _incorrect,_ as I run into
```
ifeq (,)
/bin/sh: -c: line 0: syntax error near unexpected token `,'
/bin/sh: -c: line 0: `ifeq (,)'
```
when trying to compile my crummy code.
If someone who knows what they're doing wants to implement "feature detection" for the `base64` command, by all means, go ahead.

**Comment by Vogtinator** [comment]
> ifeq ($(shell base64 -w 2>/dev/null),) is probably more correct than ifeq ($(base64 -w 2>/dev/null),)

Yeah, I don't think `$(base64 ...)` has the desired effect as there's no `base64` make builitin.

>  - but apparently still incorrect, as I run into

That means it was run as part of the rule command passed to `sh`, maybe just drop the indentation? See https://www.gnu.org/software/make/manual/html_node/Recipe-Syntax.html#Recipe-Syntax. If that doesn't work, you could do the conditional evaluation as part of the shell recipe (though everything has to be on a single logical line).

> I can at least say with confidence that the toolchain compiles successfully on macOS, once -a is removed from tar.

For autodetection of the compression algorithm, but that might  not . Just to be sure, you didn't change anything other than `-xaf` -> `-xf`, right?

**Comment by fridtjof** [comment]
> > I can at least say with confidence that the toolchain compiles successfully on macOS, once -a is removed from tar.
>
> For autodetection of the compression algorithm, but that might not . Just to be sure, you didn't change anything other than -xaf -> -xf, right?

Apparently, the tar macOS uses (BSD tar) always auto detects compression formats without any special parameters:

https://www.freebsd.org/cgi/man.cgi?tar(1)

CTRL+F for "compression automatically when reading archives"

**Comment by Vogtinator** [comment]
> I can also confirm removing -a does the trick on macOS.

Even better, so both BSD and GNU tar are happy with just `-xf`. I dropped the `-a` now.

**Comment by Kakise** [comment]
> > ifeq ($(shell base64 -w 2>/dev/null),) is probably more correct than ifeq ($(base64 -w 2>/dev/null),)
> 
> Yeah, I don't think `$(base64 ...)` has the desired effect as there's no `base64` make builitin.
> 
> > * but apparently still incorrect, as I run into
> 
> That means it was run as part of the rule command passed to `sh`, maybe just drop the indentation? See https://www.gnu.org/software/make/manual/html_node/Recipe-Syntax.html#Recipe-Syntax. If that doesn't work, you could do the conditional evaluation as part of the shell recipe (though everything has to be on a single logical line).
> 
> > I can at least say with confidence that the toolchain compiles successfully on macOS, once -a is removed from tar.
> 
> For autodetection of the compression algorithm, but that might not . Just to be sure, you didn't change anything other than `-xaf` -> `-xf`, right?

This: 
```
BASE64 := base64 -w0
ifeq ($(shell base64 -w 2> /dev/null),)
    BASE64 := base64
endif```

works perfectly on macOS

**Comment by N0ury** [comment]
> > > ifeq ($(shell base64 -w 2>/dev/null),) is probably more correct than ifeq ($(base64 -w 2>/dev/null),)
> > 
> > 
> > Yeah, I don't think `$(base64 ...)` has the desired effect as there's no `base64` make builitin.
> > > * but apparently still incorrect, as I run into
> > 
> > 
> > That means it was run as part of the rule command passed to `sh`, maybe just drop the indentation? See https://www.gnu.org/software/make/manual/html_node/Recipe-Syntax.html#Recipe-Syntax. If that doesn't work, you could do the conditional evaluation as part of the shell recipe (though everything has to be on a single logical line).
> > > I can at least say with confidence that the toolchain compiles successfully on macOS, once -a is removed from tar.
> > 
> > 
> > For autodetection of the compression algorithm, but that might not . Just to be sure, you didn't change anything other than `-xaf` -> `-xf`, right?
> 
> This:
> 
> ```
> BASE64 := base64 -w0
> ifeq ($(shell base64 -w 2> /dev/null),)
>     BASE64 := base64
> endif
> ```
> 
> works perfectly on macOS

This is probably not native base64.

```
base64 -w0
base64: invalid option -- w
Usage:	base64 [-hvDd] [-b num] [-i in_file] [-o out_file]
  -h, --help     display this message
  -Dd, --decode   decodes input
  -b, --break    break encoded string into num character lines
  -i, --input    input file (default: "-" for stdin)
  -o, --output   output file (default: "-" for stdout)
```

**Comment by Kakise** [comment]
> > > > ifeq ($(shell base64 -w 2>/dev/null),) is probably more correct than ifeq ($(base64 -w 2>/dev/null),)
> > > 
> > > 
> > > Yeah, I don't think `$(base64 ...)` has the desired effect as there's no `base64` make builitin.
> > > > * but apparently still incorrect, as I run into
> > > 
> > > 
> > > That means it was run as part of the rule command passed to `sh`, maybe just drop the indentation? See https://www.gnu.org/software/make/manual/html_node/Recipe-Syntax.html#Recipe-Syntax. If that doesn't work, you could do the conditional evaluation as part of the shell recipe (though everything has to be on a single logical line).
> > > > I can at least say with confidence that the toolchain compiles successfully on macOS, once -a is removed from tar.
> > > 
> > > 
> > > For autodetection of the compression algorithm, but that might not . Just to be sure, you didn't change anything other than `-xaf` -> `-xf`, right?
> > 
> > 
> > This:
> > ```
> > BASE64 := base64 -w0
> > ifeq ($(shell base64 -w 2> /dev/null),)
> >     BASE64 := base64
> > endif
> > ```
> > 
> > 
> > works perfectly on macOS
> 
> This is probably not native base64.
> 
> ```
> base64 -w0
> base64: invalid option -- w
> Usage:	base64 [-hvDd] [-b num] [-i in_file] [-o out_file]
>   -h, --help     display this message
>   -Dd, --decode   decodes input
>   -b, --break    break encoded string into num character lines
>   -i, --input    input file (default: "-" for stdin)
>   -o, --output   output file (default: "-" for stdout)
> ```

Tried with a fresh macOS 10.15.7 install.
Can you try this patch https://gist.github.com/Kakise/637740402370369366cd95e125e64ccb ? It should work

**Comment by Vogtinator** [comment]
> Can you try this patch https://gist.github.com/Kakise/637740402370369366cd95e125e64ccb ? It should work

That would break build on systems where `-w0` is necessary, here the condition is also true.

**Comment by Kakise** [comment]
> > Can you try this patch https://gist.github.com/Kakise/637740402370369366cd95e125e64ccb ? It should work
> 
> That would break build on systems where `-w0` is necessary, here the condition is also true.

How would it break it ? The "BASE64" variable is defaulted to `base64 -w0`and is changed only if the -w option is not implemented. Or am I missing something ? Genuinely asking because I'm bad with makefiles lol

**Comment by Vogtinator** [comment]
> > > Can you try this patch https://gist.github.com/Kakise/637740402370369366cd95e125e64ccb ? It should work
> > 
> > 
> > That would break build on systems where `-w0` is necessary, here the condition is also true.
> 
> How would it break it ? The "BASE64" variable is defaulted to `base64 -w0`and is changed only if the -w option is not implemented. Or am I missing something ? Genuinely asking because I'm bad with makefiles lol

The detection doesn't work. Here the `w` option is implemented and `ifeq ($(shell base64 -w 2> /dev/null),)` is still true.

**Comment by unilock** [comment]
The latest hastily-pushed commit *should* work. I didn't have time to test in the context of building Ndless.

**Comment by Vogtinator** [comment]
> The latest hastily-pushed commit _should_ work. I didn't have time to test in the context of building Ndless.

It doesn't seem to work. `base64 -w` exits with an error here:

```
> base64 -w
base64: option requires an argument -- 'w'
Try 'base64 --help' for more information.
> echo $?
1
```

**Comment by unilock** [comment]
> It doesn't seem to work. `base64 -w` exits with an error here:

Ah, that's because `-w` expects a valid wrap size. Try the new commit.
Only tested on Linux; not macOS. But it "should" work on both.

**Comment by Vogtinator** [comment]
> Only tested on Linux; not macOS. But it "should" work on both.

Yeah, confirmed to work here. The main point of this is to fix it for macOS, so this really should be tested there before merging.

**Comment by unilock** [comment]
Seems like it's working on my end.
Screen Shot 2020-10-25 at 10 32 09 AM

**Comment by Vogtinator** [comment]
Ok.

The use of `$?` with pipes can be a bit weird, so could you optimize the pipe away, like `base64 -w0 >/dev/null 2>&1 </dev/null`?

**Comment by unilock** [comment]
Done. I can confirm that works on both Linux and macOS.

**Review by Vogtinator — APPROVED** [review]

**Pull request merged**